org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java

```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
+     * @param isSparse - if true the integer numbers in inst-block are added in the Int set in universe, otherwise only the range is generated. 
+     * @throws ErrorSyntax
+     */
+    public CommandScope(Pos pos, Sig sig, boolean isExact, int startingScope, int endingScope, int increment, boolean isSparse) throws ErrorSyntax {
+        this(pos,sig,isExact,startingScope,endingScope,increment,new ArrayList(),0,isSparse);
+    }
+    
+    public CommandScope(Pos pos, Sig sig, boolean isExact, 
+            int startingScope, int endingScope, int increment, 
+            List/*<ExprVar>*/ atoms, int LIndex) throws ErrorSyntax {
+        this(pos,sig,isExact,startingScope,endingScope, increment, atoms,LIndex,false);
+    }
+
+    /**
+     * Create a commandscope for a sig. 
+     * If the atoms list is empty, it no sig should be created.  
+     * If not isExact and LIndex = 0 and atoms.size() > 0 then hasUpper and notLower
+     * If not isExact and LIndex > 0 and atoms.size() = LIndex then HasLower
+     * If not isExact and LIndex > 0 and atoms.size() > LIndex the hasLower and hasUpper, means a range.
+     * If isExact then LIndex = atoms.size() = LIndex
+     * @param pos
+     * @param sig
+     * @param isExact
+     * @param startingScope
+     * @param endingScope
+     * @param increment
+     * @param atoms
+     * @param lower
+     * @param hasUpper
+     * @throws ErrorSyntax
+     */
+    public CommandScope(Pos pos, Sig sig, boolean isExact, 
+            int startingScope, int endingScope, int increment, 
+            List/*<ExprVar>*/ atoms, int LIndex,boolean isSparse) throws ErrorSyntax {
+        if (pos == null) pos = Pos.UNKNOWN;
+        if (sig == null) throw new NullPointerException();
+        if (startingScope < 0) throw new ErrorSyntax(pos, "Sig "+sig+" cannot have a negative starting scope ("+startingScope+")");
+        if (endingScope < 0) throw new ErrorSyntax(pos, "Sig "+sig+" cannot have a negative ending scope ("+endingScope+")");
+        if (endingScope < startingScope) throw new ErrorSyntax(pos, "Sig "+sig+" cannot have an ending scope ("+endingScope+") smaller than its starting scope ("+startingScope+")");
+        if (startingScope == endingScope) increment = 1;
+        if (increment < 1) throw new ErrorSyntax(pos, "Sig "+sig+"'s increment value cannot be "+increment+".\nThe increment must be 1 or greater.");
         this.pos = pos;
         this.sig = sig;
         this.isExact = isExact;
-        this.startingScope = startingScope;
-        this.endingScope = endingScope;
+        this.hasUpper = !isExact & LIndex < atoms.size();
+        this.hasLower = !isExact & LIndex>0;
+        this.pAtomsLowerLastIndex=LIndex;
         this.increment = increment;
+        this.pFields = new ArrayList<List<Expr>>();
+        this.isSparse = isSparse;
+        boolean isRelation = false;
+
+        List<List<Expr>> tmpPField = new ArrayList<List<Expr>>();
+        if(atoms != null  && atoms.size() > 0){
```

Comment:
needs space between `if` and and open paren; same for `for` below

Reply:
fixed